cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java

```diff
@@ -87,7 +90,7 @@
   private final Configuration hConf;
   private final LocationFactory locationFactory;
   private final MetricsCollectionService metricsCollectionService;
-  private final DatasetFramework datasetFramework;
+  private DatasetFramework datasetFramework;
```

Comment:
Better leave this field final and have it representing the original datasetFramework that get injected through constructor. In the run() method, just have a local variable for the optionally new ds framework:

```java
DatasetFramework workflowDSFramework = datasetFramework;
if (arguments.hasOption(...)) {
  workflowDSFramework = new WorkflowDatasetFramework(...);
}
```

The reason is if anyone uses the datasetFramework field before the rewrite, it will be wrong.